Replicas annotation

kow3ns opened the pull request:
This PR adds an annotation to Pods that are members of a Stateful Set. The annotation is "pod.alpha.kubernetes.io/replicas" and its value is set to the string representation of the StatefulSet.Spec.Replicas value for its parent container. The purpose of this addition is to allow for the Pod's entry point, its init container(s), and/or its side cars to consume the number of replicas via Downward API volumes. This capability, used in conjunction with the KubeDNS, will allow a Pod to determine the full membership of the Stateful Set as intended by the administrator. This may differ from the actual membership due to node failures, network partitions, etc, but it will still be useful for configuring applications that require the full intended membership of the Stateful Set (e.g ZooKeeper, Galera) and for providing entry points to applications that require a subset of the previously launched active members (e.g. Cassandra, Riak).

``` release-note
The annotation "pod.alpha.kubernetes.io/replicas" on Stateful Sets is now provided by the Stateful Sets 
controller. This annotation tells the Pods how many replicas have been requested by the user. It should only be consumed via Downwards API Volumes.
```
<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35987)
<!-- Reviewable:end -->

Commit:
Adding replicas annotation to Pods in a StatefulSet

Commit:
Updates tests and adds comments

Commit:
Merge branch 'master' of https://github.com/kubernetes/kubernetes into replicas_annotation
pulling master in to sync prior to pushing branch to remotes

Commit:
Updates tests and adds comments

Commit:
Fixes gofmt

kow3ns commented:
@k8s-bot unit test this

kow3ns commented:
@smarterclayton 
@foxish

foxish reviewed (changes requested):
Further changes: removing the unused copyPetID function as we discussed offline.

foxish commented on pkg/controller/petset/pet.go:
```diff
@@ -63,6 +67,65 @@ type pcb struct {
+		return false, errors.New("nil source supplied")
+	} else if desired == nil {
+		return false, errors.New("nil target supplied")
+	}
```
The presence of this code seems to imply that we expect that sometimes it will be called this way, but as I understand it, this shouldn't happen. If we had to have it, I think we could just check that all 3 are not nil in a single statement and return an error.

foxish replied:
We should also remove `copyPetID` from `identity_mappers.go`. This does seem like a better place for the update function. And can we mention in the comment that the pod parameter is updated in-place?

kow3ns replied:
So the copyPetID method had these checks, and that's why they are still included. I don't think they should ever be necessary, but I didn't want to remove them as they may have a purpose that I  haven't yet seen.

kow3ns replied:
I'd rather not move update though. Its a package private method that is only invoked from the file that contains it. I don't think that it provides better clarity to move it, and I don't believe it will be useful to invoke it from other functions in the package. The only reason not to inline it is because the exported Update method would be a bit long.

smarterclayton commented:
How are you ensuring all the members of the set see the latest size correctly?  This isn't exactly the approach I would have expected - distributing the scale to each individual pet doesn't actually ensure you can get a consistent view of the total scale.  Only strongly consistent reads to the core set object can provide that.

smarterclayton commented:
Working through a scenario:

1. Create a SS of size 5
2. Pods 1-4 are created
3. User resizes to 3
3. Node for pod 1 goes down and pod is cleaned up
4. Pod 1 is created with annotation 3, pods 2-4 have annotation 5
5. Pod 4 is deleted
6. When are pods 2-3 fixed?

kow3ns commented:
@smarterclayton Lets walk through what's been implemented

I claim that, in the StatefulSetController, all of the StatefulSet's Pods have their replicas annotation updated and synchronized every time the StatefulSet.Sync.Replicas field is modified.
1. The StatefulSetController manages updates to the StatefulSet in [Sync](https://github.com/kow3ns/kubernetes/blob/replicas_annotation/pkg/controller/petset/pet_set.go#L282-L320).
2. Sync calls [syncStatefulSet](https://github.com/kow3ns/kubernetes/blob/replicas_annotation/pkg/controller/petset/pet_set.go#L323-L370) which synchronizes the StatefulSet and all of its pods.
3. This method invokes [Sync](https://github.com/kow3ns/kubernetes/blob/replicas_annotation/pkg/controller/petset/pet.go#L157-L199) on petSyncer which invokes [Update](https://github.com/kow3ns/kubernetes/blob/replicas_annotation/pkg/controller/petset/pet.go#L281-L301) which, in turn, invokes [update](https://github.com/kow3ns/kubernetes/blob/replicas_annotation/pkg/controller/petset/pet.go#L97-L127) . This updates any of the Pod's incorrect identifiers and its replicas annotation.

I also claim that, provided that kubelet is not partitioned from the API server, the Downward API volumes will receive the update atomically.
1. Every time kubelet updates a Pod there it waits for the Pod's volumes to [attach and mount](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1465-L1470l), and, as described in the comments of the [volume manager code](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/volume_manager.go#L327-L365) that implements this, the Downward API volumes will be updated with correct annotation values every time the kubelet syncs the Pod. 
2.Looking at the implementation of the [downwardAPIPlugin](https://github.com/kow3ns/kubernetes/blob/replicas_annotation/pkg/controller/petset/pet.go#L97-L127), there is a relatively clever implementation of file system mutation that provides atomic updates to the files in the Downward API volume.

So in the example above when the user resizes to 3 in (3), Pods 1-5 will all have their annotations updated, and 3-4 will additionally have their deletion timestamp set. Pods 1-5 will be able to observe the modification by watching their Downward API volumes (should they choose to mount one) with inotify or fschange. 

I'm not entirely sure what you mean by "strongly consistent" here, but based on the above I would claim that observers of the Downward API volume will observe atomic, linear (the sync loop in kubelet provides the linearization point), eventual consistent (all observers will eventually agree on the same value after mutations stop, but they may not observe all intermediate values) updates. You can think of the Downward API volume as an eventually consistent register.

Commit:
Merge branch 'master' of https://github.com/kubernetes/kubernetes into replicas_annotation
merging latest

kow3ns commented on pkg/controller/petset/pet.go:
```diff
@@ -63,6 +67,65 @@ type pcb struct {
+			pod.Name, replicaStr, StatefulSetReplicasAnnotation)
+		return -1
+	} else {
+		return int32(replicas64)
```
I'll change it to replicas, replicas64 was to indicate the return type of the strconv method prior to the explicit int32 cast in the return statement.

kow3ns commented on pkg/controller/petset/pet.go:
```diff
@@ -45,6 +46,9 @@ const (
 	// TODO: Replace this with init container status.
 	StatefulSetInitAnnotation = "pod.alpha.kubernetes.io/initialized"
+	//StatefulSetReplicasAnnotation is used to communicate the current number of replicas
+	//configured for the Stateful Set to the StatefulSet's Pods
```
done

Commit:
Changes //[comment] to // [comment]
Moves errors import declaration to alphabetical order in the section used for builtin imports
Remove copyPetID
Changes setReplicas and getReplicas to setReplicasAnnotation and getReplicasAnnotation for clarity
Changes replicas64 to replicas in getReplicasAnnotation

smarterclayton commented:
> I also claim that, provided that kubelet is not partitioned from the API server, the Downward API volumes will receive the update atomically.

Even if they aren't partitioned, they can be arbitrarily delayed.  I don't know that we're worried about the non-partition use case when we talk about these sorts of decisions - eventually consistency isn't enough to forum a quorum.

If the point of this change is to indicate the size of a quorum, you cannot guarantee that all pods receive that quorum change at the same time or at all.  So encouraging someone to use it for that would be unsafe (an eventually consistent register for use in quorum decisions isn't safe).

How will you prevent an arbitrarily delayed annotation update from leading to a split quorum?

1. Create StatefulSet size 2
2. Create pod 1 which then starts and reports ready
3. Create pod 2 which then starts and reports ready
4. Pods 1 and 2 forum quorum size 2.
5. Change set size to 5
6. Update pod 1 and pod 2 annotations, delivery arbitrarily delayed
7. Create pod 3 with size 5
8. Create pod 4 with size 5
9. Create pod 5 with size 5
10. Pods 3, 4, 5 form quorum of size 3/5

You would have to guarantee the process has received replica size changes in an ordered fashion, which might be possible if you included that info in pod status.  But we have no mechanism for that.

Commit:
fix gofmt

Commit:
fix gofmt

smarterclayton commented:
The StatefulSet controller wouldn't be able to take step 7 until it knows that pods 1 and 2 have acknowledged and used the annotation from step 6 in a strongly consistent fashion.

kow3ns commented:
@smarterclayton I disagree, lets consider that scenario.

1. In order to have a quorum size of 2 you would have to have 3 processes in the cluster. That is, unless your application supports creating a process weighted quorum. Lets ignore weighted quorums for now because most ACID systems that use ZAB, multi-Paxos, Raft, 2PC, or 3PC for consistency (read linearization of single operations) do not support this feature.

2.When you create 3 Pods they will be configured to form a cluster with 3 members. They will add the network identities of all three members and launch their processes which will use an appropriate leader election algorithm to decide a leader.

3. When you scale from 3 to 5, Pods 3 - 5 will configure their membership with 1,2,3,4, and 5. Processes 4 and 5 can not become leaders because they expect a quorum of 3, not 2. Also, depending on the leader election algorithm, they likely lag to far behind the current leader to be elected. So arbitrarily delayed delivery of the notification to 1, 2, and 3 is not an issue for causing a write loss due to a sloppy quorum.

Commit:
change getReplicas to getReplicasAnnotation in unit test

smarterclayton commented:
> In order to have a quorum size of 2 you would have to have 3 processes in the cluster. That is, unless your application supports creating a process weighted quorum. Lets ignore weighted quorums for now because most ACID systems that use ZAB, multi-Paxos, Raft, 2PC, or 3PC for consistency (read linearization of single operations) do not support this feature.

We're not talking about real quorum systems.  We're talking about elasti search (which allows quorums of 2), or JGroups, or Cassandra, or a hundred other systems that don't have Raft or PAXOS at the core.  That's who StatefulSets are targeted for.  I'm skeptical that the relatively unskilled operator can reason about the less safe systems without a lot of hand holding.

> When you create 3 Pods they will be configured to form a cluster with 3 members. They will add the network identities of all three members and launch their processes which will use an appropriate leader election algorithm to decide a leader.

Take my example, increase from 2 initial to 3 initial, then go from 3 to 7.  You still have 2 sides, but you're right, if the person who wrote the StatefulSet was savvy they would have had 4-7 only talk to pet 1 in order to join.  But many consensus systems we will use in StatefulSets may not work that way - they only accept config files that say (here are my seven members).  And throw in a relatively non-distributed systems operator (who has never heard of PAXOS or strong consensus but may have read and understood the docs of random system X) who says "oh, I can trust that is 7" who is scripting a config file in bash.  That operator will break, and they'll break because we gave them a tool they didn't understand the implications of.

You're arguing that because a sophisticated user may be able to reason about replicas correctly, we should add this annotation.  I'm arguing that our target audience has to be able to reason about the tools we give them in a safe way (fail safe).  We need a lot of guidance in order to make this annotation safe, and the naive user is very likely to be burned.

If you can come up with a succinct way of explaining when this annotation would be unsafe to use for the items in this list (https://docs.google.com/spreadsheets/d/1DDx0O35dSx73v5oFZbq17gN0SKTIDsyx3qn-gvk7Vqc/edit#gid=0) and be able to communicate that as part of this, I'd be more inclined to experiment with it.

kow3ns commented:
@smarterclayton What you're saying make a lot of sense, and I think I should try to back up and communicate exactly what I'm trying to solve here. 

In [Brewer's key note at PODC](https://people.eecs.berkeley.edu/~brewer/cs262b-2004/PODC-keynote.pdf), when his conjecture, aka the CAP theorem was born, he proposed that distributed systems lie along a spectrum between ACID and BASE, and your point that we need to support all the systems on that spectrum is well taken.

Systems like Cassandra, Riak, and Elastic Search fall more toward the BASE side. When you're talking about a quorum here you're really dealing with configurable consistency in a BASE system. That is read and write quorums and the number of configured replicas can scale independently of the number of processes in the system (at least for Cassandra and Riak).

Another nice property that BASE systems tend to have is that you can add a new process to the cluster by providing it with a set of active entry points to discover its peers, and they will, in general, do the rest of the heavy lifting for you.

For these guys, using KubeDNS to discover the active processes in the cluster and then configuring a newly joining process is a good solution, and its a pattern that seems to be implemented quite a bit in the community.

I'm concerned more about how we will deploy and scale systems that are on the ACID side of the spectrum. For these guys, a quorum has different implications. Applications that require election of a stable leader [can't do so in an anonymous ring](http://dl.acm.org/citation.cfm?id=804655). This means that for the ZooKeepers and the etcds of the world, they need to have their membership explicitly configured.

Right now the examples for these systems are using KubeDNS in much the same way as the BASE systems, and that introduces many scenarios where writes could be lost due to leader elections being triggered when the cluster is only partially configured with the number of processes intended by the operator. For example, in order to launch an ensemble of 3 ZK servers using the latest stable release, I have to do three rolling restarts https://github.com/kubernetes/contrib/blob/master/pets/zookeeper/init/on-change.sh. We've also found issues using this method with Percona where process failures led to permanent [split brain](https://github.com/kubernetes/kubernetes/issues/33938).

All I'm trying to do is convey the current number of replicas (as configured by the operator) to the Pod's initialization infrastructure so I can reconstruct the full "intended" membership of the cluster and avoid these scenarios.

I also agree that consumers of this annotation need to be well aware of the implications and considerations surrounding, and it is my intention to provide examples and detailed documentation over the next month between feature freeze prior to shipping 1.5 GA.

If the community finds this feature not to be of use or or to be confusing, either prior to GA or post (it is alpha), or if the documentation and examples are found to be lacking, then we can and should deprecate it.

smarterclayton commented:
> We've also found issues using this method with Percona where process failures led to permanent split brain.

Yes, that's why the process of managing StatefulSets must be as explicit as possible in leading users into dangerous situations, by only allowing dangerous usage when users opt in, not out.  In this case, the proposal is dangerous, because as mentioned above it leads to users believing they can safely rely on that annotation without realizing why it is dangerous.  Since that violates the design goals for StatefulSets, I don't think this as proposed will work.  Fortunately, I do think we can simplify this proposal to the point where it's safe by default and something you have a chance of landing before code freeze on monday (no promises though, there's a lot of other stuff in the queue and this is very late).

The fundamental problem is that you are giving all stateful sets an eventually consistent register.  Rather than doing that, we should give them a linearizable one.  We can do that by:

1. Limiting the number of registers down to 1 by default (prevents users from assuming things about cross node consistency that they shouldn't), and most software is better off having a strongly consistent initializer with followers joining later.
2. Make the register strongly consistent by requiring a full confirmation that the node process has the value.

The algorithm for that sounds like

1. Set the annotation on the first instance only (member 0)
2. When replica set size is changed, delete the first member, wait for clean shutdown, and recreate it with the new annotation.
3. Proceed with other member creation / management.

I would probably prefer the annotation to be opt in, because we explicitly make all APIs that are communicated to clients explicit, so the annotation would need to be set in the pod template for it to be set (if omitted, it should not be set, and step 2 above is not necessary).  If you can make those changes by noon monday I should be able to do a review before the deadline.  If not, this is going to have to wait for 1.6.

We can discuss more flexible options in the 1.6 timeframe.

smarterclayton commented:
Also, I'll note that you could today, without any code changes, put the replicas annotation on the template and get the desired outcome you seek for prototyping (at the cost of naive scale doing nothing).  A more general solution for the eventually consistent register would be to extend the downward API to support injecting the set size into the pod template, which when we add update logic to stateful sets would allow you to manage that seemlessly without special logic.  

I'm generally supportive of the more restrictive solution proposed above because it is more useful for all software that needs to form a quorum and fits the intent of the set to make config changes serializable in a useful way, but you at least have a path in 1.5 without any changes.

kow3ns commented:
@smarterclayton What you are proposing is effectively a managed rolling restart. I implemented this for multiple Mesos frameworks. [DC/OS Cassandra Framework](https://github.com/mesosphere/dcos-cassandra-service) is one example implementation.

Managed rolling restart is a powerful and useful feature, and I'd be happy to make such a contribution to K8s Stateful Sets. However, when thinking about how this would work, in this context, for an operator at 3 AM, I imagine too many scenarios that would provide a surprising user experience.

For instance, when a user patches a Stateful Set's replicas, and they are expecting more instances of the same process (e.g Scaling up for a read fan out issue due to an unexpected demand). If they instead find that they have triggered a rolling restart to their already deployed Pods that failed in the middle and left the existing cluster in an unexpected and erroneous state with less read capacity,  mitigation would turn to escalation.

If you're interested in a rolling restart feature, I'd like to work with you on fleshing that out for the 1.6 time frame, and I think we could start with deciding what use cases should be supported (e.g. software updates, cluster configuration updates, ...), what events should trigger the restarts, and what semantics the restarts should guarantee.

For what I'm trying to accomplish in the 1.5 time frame, this PR will allow me to more easily deploy several stateful applications of  interest, that make use of the provided semantics of Stateful Sets, as examples in the repository, and mitigate the current use of KubeDNS to perform a function for which it is ill suited. My hope is that the programs and helm charts that I provide will allow users to deploy these systems without concerning themselves with this annotation or its semantics.

smarterclayton commented:
It's not.  Updates (or at least a design) are currently planned for 1.6 -
I'm just giving you a path to get limited set size updates delivered to you
in a safe way without taking the step to full rolling updates (which aren't
going in without a prooosal).

If you don't have a short term use for a mutable, set-safe approach for
size management, I'm ok to punt altogether for 1.5.

If you need safe initial quorum, set an annotation / env var with your
bootstrap set size, have all members attempt to join the quorum containing
member-0, and avoid calculating a set on your own.  You are guaranteed not
to have more than one member-0 on the cluster at any time, so that gives
you a strongly consistent set you can base your bootstrap on.  Telling your
users not to scale until an initial quorum is reached is perfectly
acceptable for 1.5.

If there's a goal you can't reach in 1.5 that is not a known, already
acknowledged gap and that does not lead to unsafe reasoning for some users,
I'd recommend exploring a babysitter on top of Kube (some variant is
inevitable).  As is this change doesn't have the necessary justification to
do a no-design, last minute merge, and we don't rush things like this.  I'm
happy to engage further in 1.6 planning to help you make BASE systems
easier to deploy.

On Nov 6, 2016, at 2:54 AM, Kenneth Owens <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> What you are proposing
is effectively a managed rolling restart. I implemented this for multiple
Mesos frameworks. DC/OS Cassandra Framework
<https://github.com/mesosphere/dcos-cassandra-service> is one example
implementation.

Managed rolling restart is a powerful and useful feature, and I'd be happy
to make such a contribution to K8s Stateful Sets. However, when thinking
about how this would work, in this context, for an operator at 3 AM, I
imagine too many scenarios that would provide a surprising user experience.

For instance, when a user patches a Stateful Set's replicas, and they are
expecting more instances of the same process (e.g Scaling up for a read fan
out issue due to an unexpected demand). If they instead find that they have
triggered a rolling restart to their already deployed Pods that failed in
the middle and left the existing cluster in an unexpected and erroneous
state with less read capacity, mitigation would turn to escalation.

If you're interested in a rolling restart feature, I'd like to work with
you on fleshing that out for the 1.6 time frame, and I think we could start
with deciding what use cases should be supported (e.g. software updates,
cluster configuration updates, ...), what events should trigger the
restarts, and what semantics the restarts should guarantee.

For what I'm trying to accomplish in the 1.5 time frame, this PR will allow
me to more easily deploy several stateful applications of interest, that
make use of the provided semantics of Stateful Sets, as examples in the
repository, and mitigate the current use of KubeDNS to perform a function
for which it is ill suited. My hope is that the programs and helm charts
that I provide will allow users to deploy these systems without concerning
themselves with this annotation or its semantics.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/35987#issuecomment-258665936>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_pxtbAzRkUUe2Bi0YqUKeMkhkcyemks5q7Ye7gaJpZM4KmbiW>
.
